skxray/core/speckle.py

```diff
@@ -0,0 +1,288 @@
+
+            while processing:
+                if not track_level[level]:
+                    track_level[level] = 1
+                    processing = 0
+                else:
+                    prev = 1 + (cur[level - 1] - 2) % timebin_num
+                    cur[level] = 1 + cur[level] % timebin_num
+
+                    buf[level, cur[level]-1] = (buf[level-1,
+                                                    prev-1] +
+                                                buf[level-1,
+                                                    cur[level - 1] - 1])
+                    track_level[level] = 0
+
+                    _process(num_roi, level, cur[level]-1, buf, img_per_level,
+                             labels, max_cts, bin_edges[level, 0], prob_k,
+                             prob_k_pow)
+                    level += 1
+                    # Checking whether there is next level for processing
+                    processing = level < num_times
+
+            prob_k_all += (prob_k - prob_k_all)/(i + 1)
+            prob_k_pow_all += (prob_k_pow - prob_k_pow_all)/(i + 1)
+
+    prob_k_std_dev = np.power((prob_k_pow_all -
+                               np.power(prob_k_all, 2)), .5)
+
+    logger.info("Processing time for XSVS took %s seconds."
+                "", (time.time() - start_time))
+    return prob_k_all, prob_k_std_dev
+
+
+def _process(num_roi, level, buf_no, buf, img_per_level, labels,
+             max_cts, bin_edges, prob_k, prob_k_pow):
+    """
+    Internal helper function. This modifies inputs in place.
+
+    This helper function calculate probability of detecting speckles for
+    each integration time.
+
+    Parameters
+    ----------
+    num_roi : int
+        number of ROI's
+    level : int
+        current time level(integration time)
+    buf_no : int
+        current buffer number
+    buf : array
+        image data array to use for XSVS
+    img_per_level : int
+        to track how many images processed in each level
+    labels : array
+        labels of the required region of interests(ROI's)
+    max_cts: int
+        maximum pixel count
+    bin_edges : array
+        bin edges for each integration times and each ROI
+    prob_k : array
+        probability density of detecting speckles
+    prob_k_pow : array
+        squares of probability density of detecting speckles
+    """
+    img_per_level[level] += 1
+    u_labels = list(np.unique(labels))
+
+    for j in range(num_roi):
+        roi_data = buf[level, buf_no][labels == u_labels[j]]
+
+        spe_hist, bin_edges = np.histogram(roi_data, bins=bin_edges,
+                                           density=True)
+
+        prob_k[level, j] += (np.nan_to_num(spe_hist) -
+                             prob_k[level, j])/(img_per_level[level])
+
+        prob_k_pow[level, j] += (np.power(np.nan_to_num(spe_hist), 2) -
+                                 prob_k_pow[level, j])/(img_per_level[level])
+
+    return  # modifies arguments in place!
```

Comment:
You do not need to add a `return` or `return None` at the very end of a function as Python returns None for you.